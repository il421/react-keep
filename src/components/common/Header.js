import React from 'react';
import { connect } from 'react-redux';

import { startLogout } from '../../actions/auth';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import Search from './Search';

export const Header = ({ startLogout, auth, showSidebar }) => {

  return (
    <header className="header" onClick={(evt) => evt.stopPropagation()}>
      <div className="content-container">
        <div className="header__content">
          <div className="header__sidebar">

            <button className="button button--sidenav" onClick={() => showSidebar()}>
              <FontAwesomeIcon icon="bars" size="2x" />
            </button>

          </div>
          <div>
            <h1 className="header__title">Keep Me</h1>
          </div>
          <div className="header__search">
            <Search />
          </div>
          <div className="header__user-box user-box">
            <div className="user-box__photo">
              <img src={ auth.url } width={ 50 } height={ 50 } />
            </div>
            <div className="user-box__name show-for-mobile">{ auth.name }</div>
            <button className="user-box__logout button button--sidenav" onClick={ startLogout }>
              <FontAwesomeIcon icon="sign-out-alt" size="lg" />
            </button>
          </div>
        </div>
      </div>
    </header>
  );
};


const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
    url: state.url,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
