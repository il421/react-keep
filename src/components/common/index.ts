export * from "./Controllers";
export * from "./DashboardPage";
export * from "./Header";
export * from "./NotFoundPage";
export * from "./Search";
export * from "./SideBar";
