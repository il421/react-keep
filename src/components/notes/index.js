export * from "./NotesList";
export * from "./NotesItem";
export * from "./notes.types";
export * from "./TextNoteFormModal";
export * from "./ListNoteFormModal";
export * from "./NoteForm";
export * from "./options/ColorsPickerField";
