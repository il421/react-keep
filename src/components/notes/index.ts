export * from "./ConfirmDialog";
export * from "./ImageNoteFormModal";
export * from "./ListNoteFormModal";
export * from "./ListNoteFromItem";
export * from "./NoteContent";
export * from "./NoteForm";
export * from "./notes.types";
export * from "./NotesItem";
export * from "./TextNoteFormModal";
export * from "./utils";
export * from "./options/ColorsPickerField";
export * from "./options/TagsPickerField";
