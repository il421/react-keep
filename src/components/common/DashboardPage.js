import React, { useState } from 'react';

import NoteSelection from '../note/NoteSelection';
import NotesList from '../note/NotesList';
import Header from './Header';
import SideBar from './SideBar';


const DashboardPage = () => {
  let [showBar, toggleSidebarDisplay] = useState(false);

  const showSidebar = () => {
    toggleSidebarDisplay(showBar = true);
  };

  const hideSidebar = () => {
    toggleSidebarDisplay(showBar = false);
  };

  return (
    <div className="dashboard">
      <Header showSidebar={ showSidebar } />
      <NoteSelection className="content-container"/>
      <NotesList/>

      <SideBar showBar={ showBar } hideSidebar={ hideSidebar } />
    </div>
  );
};
export default DashboardPage;
