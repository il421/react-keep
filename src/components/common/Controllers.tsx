import React from "react";
import { AlignItems, JustifyContent } from "../../common";
import { FlexBox, IconButton } from "../ui-components";
import "../../styles/components/common/_controllers.scss";
import { NoteType } from "../notes";

interface Controllers extends Partial<Pick<HTMLElement, "className">> {
  isMobile?: boolean;
  openDialog: (type: NoteType) => void;
}

export const Controllers: React.FunctionComponent<Controllers> = ({
  className,
  isMobile = false,
  openDialog,
}) => {
  const handleOnClick = (type: NoteType) => () => openDialog(type);

  return (
    <div
      className={`controllers  ${
        isMobile && "controllers--mobile"
      } ${className}`}
    >
      <FlexBox
        vertical={isMobile}
        justifyContent={JustifyContent.spaceBetween}
        alignItems={AlignItems.center}
      >
        <IconButton
          className="controllers__text"
          onButtonClick={handleOnClick(NoteType.text)}
          icon="align-left"
          size="2x"
        />

        <IconButton
          className="controllers__list"
          onButtonClick={handleOnClick(NoteType.list)}
          icon="check-square"
          size="2x"
        />

        <IconButton
          className="controllers__image"
          onButtonClick={handleOnClick(NoteType.image)}
          icon="image"
          size="2x"
        />
      </FlexBox>
    </div>
  );
};
