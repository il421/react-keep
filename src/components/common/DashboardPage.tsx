import React from "react";
import { ToastContainer } from "react-toastify";
import { History, Path } from "history";
import { JustifyContent } from "../../common";
import { Header } from "./Header";
import { SideBar } from "./SideBar";
import { Controllers } from "./Controllers";
import UserFormModal from "../login/UserFormModal";
import "../../styles/components/common/_dashboard.scss";
import "../../styles/components/notes/_note-modal.scss";
import { QueryKeys } from "../../routers/Routing";
import { stringify } from "query-string";
import {
  NoteType,
  ImageNoteFormModal,
  TextNoteFormModal,
  ListNoteFormModal,
} from "../notes";
import { FlexBox } from "../ui-components";
import { NotesList } from "../notes/NotesList";

export interface DashboardPageProps {
  history: History;
}

export const onNoteSelected = (options: {
  type: NoteType;
  id: string;
  pathname: string;
  push: (path: Path) => void;
}) => {
  let key: keyof typeof QueryKeys;
  const { type, id, pathname, push } = options;
  switch (type) {
    case NoteType.text:
      key = QueryKeys.text;
      break;
    case NoteType.list:
      key = QueryKeys.list;
      break;
    default:
      key = QueryKeys.image;
  }

  const query = stringify({
    [key]: id,
  });
  push(`${pathname}?${query}`);
};

export const DashboardPage: React.FunctionComponent<DashboardPageProps> = ({
  history,
}): JSX.Element => {
  return (
    <FlexBox
      vertical={true}
      justifyContent={JustifyContent.spaceBetween}
      className="dashboard"
    >
      <Header history={history} />
      <NotesList
        onNoteSelected={(type: NoteType, id: string) =>
          onNoteSelected({
            type,
            id,
            pathname: history.location.pathname,
            push: history.push,
          })
        }
      />

      <SideBar />
      <TextNoteFormModal history={history} />
      <ListNoteFormModal history={history} />
      <ImageNoteFormModal history={history} />
      <UserFormModal />
      <Controllers
        isMobile
        openDialog={(query) =>
          history.push(`${history.location.pathname}?${query}`)
        }
      />

      <ToastContainer autoClose={2000} />
    </FlexBox>
  );
};
