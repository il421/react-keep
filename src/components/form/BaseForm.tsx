import React, { ReactNode } from "react";
import { Form, FormRenderProps } from "react-final-form";
import { FormApi, ValidationErrors } from "final-form";
import { sortArray } from "../../common";
import arrayMutators from "final-form-arrays";

/**
 * Base Form initially works with all types of notes, however also
 * can be used as a any usage form
 * @param initialValues - FormValues plus currentOption to switch between a note options
 * @param onSubmit - handle submit action, have the form values as an argument
 * @param validate - options, handle validation, return ValidationErrors
 * @param getFormActions - form actions, buttons
 * @param formClassName - optional, a number of the form components class names for styling and testing
 * @param resetAfterSubmitting - options reset after submitting
 */

export interface BaseFormProps<FormValues> {
  initialValues: FormValues;
  onSubmit: (values: FormValues) => void;
  getFormActions: (
    isDisable: boolean,
    isSubmitting: boolean,
    values: FormValues
  ) => ReactNode;
  formClassName: string;
  validate?: (values: FormValues) => ValidationErrors;
  resetAfterSubmitting?: boolean;
}

export class BaseForm<FormValues> extends React.PureComponent<
  BaseFormProps<FormValues>
> {
  private form: FormApi<FormValues> | undefined;

  componentWillUnmount() {
    this.form && this.form.pauseValidation();
  }

  render() {
    const {
      initialValues,
      formClassName,
      onSubmit,
      validate,
      getFormActions,
      resetAfterSubmitting = false,
    } = this.props;
    return (
      <Form
        initialValues={initialValues}
        onSubmit={onSubmit}
        validate={validate}
        mutators={{ sortArray, ...arrayMutators }}
      >
        {(props: FormRenderProps<FormValues>) => {
          const {
            dirty,
            values,
            submitting,
            submitSucceeded,
            form,
            handleSubmit,
          } = props;
          // reset form if form is submitted successfully
          resetAfterSubmitting && submitSucceeded && form.reset();

          if (!this.form) {
            this.form = form;
          }

          return (
            <>
              <form
                onSubmit={handleSubmit}
                className={formClassName}
                autoComplete="off"
              >
                {this.props.children}
                <div>{getFormActions(!dirty, submitting, values)}</div>
              </form>
            </>
          );
        }}
      </Form>
    );
  }
}
