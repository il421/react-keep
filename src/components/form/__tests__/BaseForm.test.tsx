import React from "react";
import { shallow, mount, ShallowWrapper, ReactWrapper } from "enzyme";
import { BaseForm, BaseFormProps } from "../BaseForm";
import { notes } from "../../../testData/notes";
import {
  NoteFormValues,
  NoteType,
  getInitialNoteFormValues,
} from "../../notes";
import { PickerColors } from "../../../common";

let props: BaseFormProps<NoteFormValues<string>>,
  wrapper: ShallowWrapper<BaseFormProps<NoteFormValues<string>>, any>;
const defaultNote: NoteFormValues<string> = {
  type: NoteType.text,
  title: "",
  content: "",
  tags: [],
  color: PickerColors.white,
};

beforeEach(() => {
  props = {
    initialValues: getInitialNoteFormValues({
      type: NoteType.text,
      currentNote: notes[0],
      defaultValues: defaultNote,
    }),
    onSubmit: jest.fn(),
    formClassName: "formClassName",
    getFormActions: jest.fn()
  };

  wrapper = shallow<BaseFormProps<NoteFormValues<string>>>(
    <BaseForm {...props} />
  );
});

test("should render BaseForm", () => {
  expect(wrapper.debug()).toMatchSnapshot();
});

test("should submit form", () => {
  const mountedForm = mount<
    ReactWrapper<BaseFormProps<NoteFormValues<string>>, any>
  >(<BaseForm {...props} />);
  expect(mountedForm.exists(".formClassName")).toBeTruthy();
  mountedForm.find("form").simulate("focus").simulate("submit");
  expect(props.onSubmit).toHaveBeenLastCalledWith(
    props.initialValues,
    expect.anything(), // formApi
    expect.anything() // options callback
  );
});
