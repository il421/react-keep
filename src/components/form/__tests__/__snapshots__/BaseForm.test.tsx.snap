// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BaseForm 1`] = `
"<ReactFinalForm initialValues={{...}} onSubmit={[Function: mockConstructor]} validate={[undefined]} mutators={{...}}>
  [function]
</ReactFinalForm>"
`;
